### PR TITLE
[#11180] Convert event-based log message to structured logs

### LINKED ISSUE #11180: Convert event-based log message to structured logs
https://github.com/TEAMMATES/teammates/blob/b23e56e666e03b60ade6649769b7971c9d782ef8/src/main/java/teammates/ui/webapi/WebApiServlet.java#L78-L81

https://github.com/TEAMMATES/teammates/blob/b23e56e666e03b60ade6649769b7971c9d782ef8/src/main/java/teammates/ui/webapi/OriginCheckFilter.java#L179-L182

https://github.com/TEAMMATES/teammates/blob/b23e56e666e03b60ade6649769b7971c9d782ef8/src/main/java/teammates/logic/api/EmailSender.java#L62-L65

These are some examples where events are logged so that they can be searched via the logs console.

Under the GAE second generation runtime integration, the system has enabled usage of [structured logging](https://cloud.google.com/logging/docs/structured-logging) which can be used to make the filtering more efficient by many orders of magnitude.

### PR DESCRIPTION
**Pull request by wkurniawan07**
Fixes #11180 

Events are logged when HTTP request is received, HTTP response is dispatched, and emails are sent. Note that this PR's purpose is just to set up basic infrastructure and example.

### CODE REVIEW

File: src/main/java/teammates/common/util/HttpRequestHelper.java [hunk 1]
```diff
@@ -21,36 +22,38 @@ private HttpRequestHelper() {
     }
 
     /**
-     * Gets the parameters of the given HTTP request as key-value (possibly multi-values) mapping string.
+     * Gets the parameters of the given HTTP request as key-value (possibly multi-values) mapping.
      */
-    public static String getRequestParametersAsString(HttpServletRequest req) {
-        return getDisplayedJsonInOneLine(req.getParameterMap());
+    public static Map<String, Object> getRequestParameters(HttpServletRequest req) {
+        Map<String, Object> params = new HashMap<>();
```
Comment (rrtheonlyone) [inline src/main/java/teammates/common/util/HttpRequestHelper.java:28]: `Map<String, Serializable>` might be a better type here I believe

File: src/main/java/teammates/common/util/HttpRequestHelper.java [hunk 2]
```diff
@@ -21,36 +22,38 @@ private HttpRequestHelper() {
...
     }
 
     /**
-     * Gets the headers of the given HTTP request as key-value (possibly multi-values) mapping string.
+     * Gets the headers of the given HTTP request as key-value (possibly multi-values) mapping.
      */
-    public static String getRequestHeadersAsString(HttpServletRequest req) {
-        Map<String, String[]> headers = new HashMap<>();
+    public static Map<String, Object> getRequestHeaders(HttpServletRequest req) {
+        Map<String, Object> headers = new HashMap<>();
```
Comment (rrtheonlyone) [inline src/main/java/teammates/common/util/HttpRequestHelper.java:43]: here too - `Map<String, Serializable>`

**Review by rrtheonlyone — APPROVED** [review]
Nice work! Looks good, just minor comments below:

One suggestion I have is to skip `StackTraceElement source = getLoggerSource();` depending on the type of logging we are doing e.g. event logging.  Although this is likely not a big deal for our use case.

File: src/main/java/teammates/ui/webapi/RequestTraceFilter.java [hunk 3]
```diff
@@ -36,9 +42,31 @@ public void init(FilterConfig filterConfig) {
...
+            return;
+        }
+
+        HttpServletRequest request = (HttpServletRequest) req;
+
+        try {
+            // Make sure that all parameters are valid UTF-8
+            request.getParameterMap();
+        } catch (RuntimeException e) {
+            if (e.getClass().getSimpleName().equals("BadMessageException")) {
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/RequestTraceFilter.java:63]: - The exception is not caught even when the catch block is defined. I'm still wondering why.
- `BadMessageException` belongs to `org.eclipse.jetty` package which is ruled to be not accessible from anywhere other than the main class. Admittedly, this is a very malleable rule.

File: src/main/java/teammates/ui/webapi/WebApiServlet.java [hunk 4]
```diff
@@ -59,63 +60,59 @@ public void doDelete(HttpServletRequest req, HttpServletResponse resp) throws IO
...
+
+            String logMessage = "%s " + RequestTracer.getTraceId() + " %s with %s in " + timeElapsed + "ms";
+            if (action == null) {
+                logMessage = String.format(logMessage, "Response", "dispatched", statusCode);
+            } else {
+                responseDetails.put("actionClass", action.getClass().getSimpleName());
+                logMessage = String.format(logMessage, action.getClass().getSimpleName(), "finished", statusCode);
+            }
+
+            log.event(LogEvent.RESPONSE_DISPATCHED, logMessage, responseDetails);
```
[the remark this answers is not in the document]
Comment (wkurniawan07) [reply src/main/java/teammates/ui/webapi/WebApiServlet.java:115]: These are not the only places where annotating the log source location is redundant, and I don't really have the bandwidth to do full investigation. Its effect to performance should be negligible at best so I don't think we should focus too much here.

**Pull request merged**